pkg/cmdline/cmdline.go

```diff
@@ -29,54 +25,29 @@ type CmdLine struct {
 	Err   error
 }
 
-var (
-	// procCmdLine package level static variable initialized once
-	once        sync.Once
-	procCmdLine CmdLine
-)
-
-func cmdLineOpener() {
-	cmdlineReader, err := os.Open("/proc/cmdline")
-	if err != nil {
-		errorMsg := fmt.Sprintf("Can't open /proc/cmdline: %v", err)
-		log.Print(errorMsg)
-		procCmdLine = CmdLine{Err: fmt.Errorf(errorMsg)}
-		return
-	}
-
-	procCmdLine = parse(cmdlineReader)
-	cmdlineReader.Close()
-}
-
 // NewCmdLine returns a populated CmdLine struct
-func NewCmdLine() CmdLine {
-	// We use cmdLineReader so tests can inject here
-	once.Do(cmdLineOpener)
-	return procCmdLine
+func NewCmdLine() *CmdLine {
+	return getCmdLine()
 }
 
 // FullCmdLine returns the full, raw cmdline string
 func FullCmdLine() string {
-	once.Do(cmdLineOpener)
-	return procCmdLine.Raw
+	return getCmdLine().Raw
 }
 
 // parse returns the current command line, trimmed
-func parse(cmdlineReader io.Reader) CmdLine {
+func parse(cmdlineReader io.Reader) *CmdLine {
+	var line CmdLine
 	raw, err := io.ReadAll(cmdlineReader)
-	line := CmdLine{}
-	if err != nil {
-		log.Printf("Can't read command line: %v", err)
+	if line.Err != nil {
 		line.Err = err
```

Comment:
I'm trying to understand how errors are handled here... I see things are setting the `Err` field of the struct, but nothing is actually reading it?   Is that field ever used? 

IIUC if parse() fails to read the `io.Reader` it sets `Err` and never sets the `AsMap` map (as expected) so any subsequent call to get flags/etc will get a nil pointer reference and barf.

It seems like setting Err in the struct instead of explicitly returning an error is a bad foot-gun for users of this package? Should we be more explicit with the error handling in parse() and it's call-hierarchy?

Reply:
I really don't like the way this was done at all. I'm afraid to change it because we don't version in u-root yet and I don't know where outside u-root this is being used. This is not my favorite u-root package.

Reply:
yeah that was a screwup while I tried to work around this awkward code. Thanks. Fixed.

Reply:
Also, thanks for the pointer on AsMap. Fixed.

Reply:
Consider that we're already changing the function signature of `NewCmdLine()` to return a `*CmdLine` instead of `CmdLine` - this means that we're already breaking consumers of this package, correct?

If so, then maybe ["no haunted graveyards"](https://www.usenix.org/sites/default/files/conference/protected-files/srecon17americas_slides_reese.pdf) applies here - let's smoke out the broken use cases and return an `err` as well while we're  already changing the signature?

Also: Maybe we should have a policy about changing/removing functions so that consumers know what to expect and we don't have to be afraid to make such changes  in the future?

Reply:
once we do proper versioning, which we will prob talk about at OSFC, this should get better.